File: solutions/day-05/day05Jokot.py
from collections import defaultdict

class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        if len(nums) == k:
            return nums
        
        dic = self.counter(nums)
        buckets = self.bucket_sort(len(nums), dic)
        top = self.get_top(buckets, k)

        return top
    
    def get_top(self, buckets, k):
        top = []
        for b in buckets[::-1]:
            for e in b:
                top.append(e)
                if len(top) == k:
                    return top
    
    def counter(self, nums):
        count = defaultdict(int)
        for n in nums:
            count[n] += 1
        
        return count

    def bucket_sort(self, size, dic):
        buckets = [[] for _ in range(size+1)]

        for key, val in dic.items():
            buckets[val].append(key)

        return buckets

File: solutions/day-05/day05Jokot.kt
class Solution {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        return getTopK(bucketSort(counter(nums), nums.size), k)
    }

    fun counter(nums: IntArray): Map<Int, Int> {
        val count = mutableMapOf<Int, Int>()

        for (n in nums) {
            count[n] = count.getOrDefault(n, 0) + 1
        }

        return count
    }

    fun bucketSort(couter: Map<Int, Int>, size: Int): List<List<Int>> {
        val buckets = MutableList(size + 1) { mutableListOf<Int>() }

        couter.forEach { n, c ->
            buckets[c].add(n)
        }

        buckets.reverse()

        return buckets
    }

    fun getTopK(buckets: List<List<Int>>, k: Int): IntArray {
        val topK = IntArray(k)
        
        var index = 0
        for (b in buckets) {
            for (n in b) {
                topK[index] = n
                index++
                if (index == k) return topK
            }
        }

        return topK
    }
}